torch/csrc/jit/ir.cpp

```diff
@@ -251,6 +255,322 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+  indent(out, level);
+  prettyPrintValue(out, lhs, pass);
+  out << " = ";
+  prettyPrintValue(out, rhs, pass);
+  out << std::endl;
+}
+
+static std::ostream& prettyPrintIf(
+  std::ostream & out,
+  const Node* node,
+  const size_t level,
+  PrettyPrintPass& pass
+) {
+  indent(out, level);
+  out << "if ";
+  const auto if_block = node->blocks()[0];
+  const auto else_block = node->blocks()[1];
+  prettyPrintValue(out, node->inputs()[0], pass);
+  out << ":" << std::endl;
+
+  // Print node contents
+  prettyPrintBlock(out, if_block, level + 1, pass);
+
+  // Print if block output
+  dualIterator<const Value*>(
+    node->outputs(),
+    if_block->return_node()->inputs(),
+    0, 0,
+    [&](const Value* node_output, const Value* return_input) {
+      prettyPrintStmt(out, node_output, return_input, level + 1, pass);
+    }
+  );
+
+  indent(out, level);
+  out << "else:" << std::endl;
+  prettyPrintBlock(out, else_block, level + 1, pass);
+  dualIterator<const Value*>(
+    node->outputs(),
+    else_block->return_node()->inputs(),
+    0, 0,
+    [&](const Value* node_output, const Value* return_input) {
+      prettyPrintStmt(out, node_output, return_input, level + 1, pass);
+    }
+  );
+
+  return out;
+}
+
+static bool isValueUsed(
+  const Value* val,
+  PrettyPrintPass& pass
+) {
+  if (pass.aliases.find(val->uniqueName()) != pass.aliases.end()) {
+    return true;
+  }
+  return val->uses().size() > 0;
+}
+
+static std::ostream& prettyPrintLoop(
+  std::ostream & out,
+  const Node* node,
+  const size_t level,
+  PrettyPrintPass& pass
+) {
+  const auto body_block = node->blocks()[0];
+
+  // Add temporaries for loop-carried dependencies
+  dualIterator<const Value*>(
+    node->inputs(),
+    body_block->param_node()->outputs(),
+    1, 0,
+    [&](const Value* node_input, const Value* param_output) {
+      if (isValueUsed(param_output, pass)) {
+        prettyPrintStmt(out, param_output, node_input, level, pass);
+      }
+    }
+  );
+
+  // Add an alias for the loop condition variable so it doesn't add an extra
+  // temporary for it
+  pass.aliases[body_block->param_node()->outputs()[0]->uniqueName()] = node->inputs()[1]->uniqueName();
+
+  // Loop header
+  indent(out, level);
+  out << "while ";
+  prettyPrintValue(out, body_block->param_node()->outputs()[0], pass);
+  out << ":";
+  out << std::endl;
+
+  // Loop body
+  prettyPrintBlock(out, body_block, level + 1, pass);
+
+  // Re-assign block outputs to inputs for next iteration
+  dualIterator<const Value*>(
+    body_block->param_node()->outputs(),
+    body_block->return_node()->inputs(),
+    0, 0,
+    [&](const Value* param_output, const Value* return_input) {
+      if (isValueUsed(param_output, pass)) {
+        prettyPrintStmt(out, param_output, return_input, level + 1, pass);
+      }
+    }
+  );
+
+  // Add assign loop-dependencies to the outputs of the prim::Loop
+  dualIterator<const Value*>(
+    node->outputs(),
+    body_block->return_node()->inputs(),
```

Comment:
Note: you can alias `node->outputs()` to `body_block->inputs()` and then you do not have to print these assignments. This is known to be safe because body_block->inputs() are not live after the block so reusing their names is ok.